### ISSUE TITLE
Zone API endpoints do not properly return zone members

### ISSUE THREAD

**Issue by elboletaire** [Sylius/Sylius#14814]
**Sylius version affected**: >= 1.11

**Description**  
Zones API is wrong. There's no proper way to get zone-members of a zone, you just get a single entry of the zone you're requesting.

**Steps to reproduce**  
- Create some zones and add some zones attached to those zones.
- Obtain an authorization bearer via `/api/v2/admin/authentication-token`
- Try to get the zone members of the zone you've been adding members to.
- Try to get the members via the API.

In my case I created "Barcelona", as a province type. Then I created some zone types assigned to Barcelona. There's no way via de API to get those zones added to Barcelona:
![imatge](https://user-images.githubusercontent.com/153305/219398157-5e0dd916-ff68-4452-8cb8-1f3c36c75a06.png)
![imatge](https://user-images.githubusercontent.com/153305/219398225-22b1a928-ea1c-4329-ba74-af58601f97e3.png)

I would expect to be able to retrieve all those numbered zones by calling to zones or zonemembers asking for Barcelona, but here are the responses I'm getting for all the available calls:

For `/api/v2/admin/zones/Barcelona`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/Zone",
	"@id": "\/api\/v2\/admin\/zones\/Barcelona",
	"@type": "Zone",
	"id": 1,
	"code": "Barcelona",
	"name": "Barcelona",
	"type": "province",
	"scope": "all",
	"members": [
		"\/api\/v2\/admin\/zone-members\/ES-BA"
	]
}
~~~

For `/api/v2/admin/zones/Barcelona/members`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/Barcelona\/members",
	"@type": "hydra:Collection",
	"hydra:member": [
		{
			"@id": "\/api\/v2\/admin\/zone-members\/ES-BA",
			"@type": "ZoneMember",
			"id": 1,
			"code": "ES-BA",
			"belongsTo": "\/api\/v2\/admin\/zones\/Barcelona"
		}
	],
	"hydra:totalItems": 1
}
~~~

For `/api/v2/admin/zone-members/ES-BA`:

~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zone-members\/ES-BA",
	"@type": "ZoneMember",
	"id": 1,
	"code": "ES-BA"
}
~~~

As you can see, none of the zone calls properly return zone members.

**Possible Solution**  
I've tried changing some of the api_platform resources and serialization files, but whatever I tried (I'm a newbie in sylius btw..) ends up with errors saying "More than one result was found for query although one row or none was expected.", so not sure how could I help here.

What's clear is this API endpoint is wrong. There should be a way to properly retrieve child zones applied to another zone.

**Comment by elboletaire** [comment]
I'm gonna close the issue because it wasn't pretty well explained, plus I now see it really works the other way, upside down (meaning there's no way I can get every zone linked to a province, because there's no way to get it... but I can know to what province a zone is linked, this is the only way the API works apparently).

What I'm trying to do is create zones for each postcode in my city. So the plan is:
- Create a city
- Create multiple zones assigned to that city
- Create postcodes assigned to zones

I expect to be able to traverse from the city down to the postcodes, but the truth is:
- I create "Barcelona" as a zone of type "province", which shows me provinces to select, in my case only one which is the one I have configured in the provinces panel: Barcelona
- I then create zones assigned to other zones, and here I create "Zone 1", "Zone 2", "Zone n", and assign them to the previously created "Barcelona" zone
- I now create the postcodes, assigned to the zones ("Zone n") I need them.

The problem:

The way the API is done, I won't be able to traverse down things, since everything is linked up, not down. So I would expect to be able to get all the zones linked to Barcelona, but there's no way. Here are the responses I'm currently getting for `zones/Barcelona/members`, `zones/Barcelona` and `zone-members/Barcelona`, respectively:

`zones/Barcelona/members`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/Barcelona\/members",
	"@type": "hydra:Collection",
	"hydra:member": [
		{
			"@id": "\/api\/v2\/admin\/zone-members\/ES-BA",
			"@type": "ZoneMember",
			"id": 1,
			"code": "ES-BA",
			"belongsTo": "\/api\/v2\/admin\/zones\/Barcelona"
		}
	],
	"hydra:totalItems": 1
}
~~~
`zones/Barcelona`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/Zone",
	"@id": "\/api\/v2\/admin\/zones\/Barcelona",
	"@type": "Zone",
	"id": 1,
	"code": "Barcelona",
	"name": "Barcelona",
	"type": "province",
	"scope": "all",
	"members": [
		"\/api\/v2\/admin\/zone-members\/ES-BA"
	]
}
~~~
`zone-members/ES-BA`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/ES-BA\/members",
	"@type": "hydra:Collection",
	"hydra:member": [],
	"hydra:totalItems": 0
}
~~~

But if I go the other way around, I can get up to Barcelona:

`zones/08032/members`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/08032\/members",
	"@type": "hydra:Collection",
	"hydra:member": [
		{
			"@id": "\/api\/v2\/admin\/zone-members\/zone-1",
			"@type": "ZoneMember",
			"id": 3,
			"code": "zone-1",
			"belongsTo": "\/api\/v2\/admin\/zones\/08032"
		}
	],
	"hydra:totalItems": 1
}
~~~
Then `zones/zone-1/members`:
~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/zone-1\/members",
	"@type": "hydra:Collection",
	"hydra:member": [
		{
			"@id": "\/api\/v2\/admin\/zone-members\/ES-BA",
			"@type": "ZoneMember",
			"id": 2,
			"code": "ES-BA",
			"belongsTo": "\/api\/v2\/admin\/zones\/zone-1"
		}
	],
	"hydra:totalItems": 1
}
~~~
And, finally 

~~~json
{
	"@context": "\/api\/v2\/contexts\/ZoneMember",
	"@id": "\/api\/v2\/admin\/zones\/ES-BA\/members",
	"@type": "hydra:Collection",
	"hydra:member": [],
	"hydra:totalItems": 0
}
~~~
Which is absolutely useless, since it does not give any information at all because it won't ever be linked with more members due to how is everything linked.

So, in conclusion... it's usable in its current state, but it should be better documented, with use cases like this one. I also think you should consider exposing publicly the zones API endpoint, since right now there's only an admin endpoint for accessing zones, which makes the entire system kinda useless if you're making a frontoffice application via the API.

Can anyone please confirm this is the expected behavior? Because as said, it's not documented at all 😞

**Comment by NoResponseMate** [comment]
Hi @elboletaire
First of all sorry for the late reply.

Zones are not exposed on the shop API because it's a pretty much backend resource, and ATM there's no sense to that. There's not much you can do with a zone alone, they are used in the checkout logic to select the proper tax values or limit the shipping methods available to the customer. Both of those things use zones, but the customer does not choose a zone, it gets resolved automatically inside the `ZoneMatcher` service using the customer's address.

Let me give you a general overview of the zone architecture.
Zones work top-down, one zone can have X members but a member is only connected to one zone.
A Zone is basically a specific geographical area. Zone Member on the other hand is just a relation between a zone and something else, in the base code that something can be either provinces, countries, or other zones, that's what the zone type is for.

Correct me if I'm wrong, but given your data, you want to create a zone using postcodes.
That's not really possible in the base implementation since there are only the 3 types mentioned above (though the use case seems quite popular and it might be worth having it in core @GSadee).

To do that you'd need to modify the code a bit.
1. Create a new resource that reflects a postcode or a postcode range
2. Add a new zone type that'll allow relating a postcode to a zone
3. Overwrite/decorate the zone matcher to take your new zone type into account
There are probably more steps, but those are the main ones.

After that's done you can feed the zone matcher an address and based on your new logic it would resolve postcodes and return fitting zones.

I don't know your exact use case, but even if you'd want to simply get a given zone on the shop API based on a postcode alone, you'd still need points 1. and 2. then create the endpoint (IMO a simple index would suffice) and add a custom APIP filter that'd, for instance, compare zone member code with passed postcode or something alike.

Either way, I'm sorry but this is not really an issue per se, and more a question about customization so I'll be closing it. Let me know if you think otherwise or have some more info/problems regarding this and I'll reopen it so we can discuss it further.

For future reference, if you have some questions about the architecture, customization, or a similar topic, you should first try reaching out on [our slack](https://sylius-devs.slack.com).

Cheers 🍻

**Comment by elboletaire** [comment]
Well, I'd say you have an issue with the documentation since, as I said, there's almost no documentation referring the zones system.
